crates/subspace-service/src/sync_from_dsn.rs

```diff
@@ -0,0 +1,191 @@
+use crate::dsn::import_blocks::import_blocks_from_dsn;
+use futures::channel::mpsc;
+use futures::{FutureExt, StreamExt};
+use sc_client_api::{BlockBackend, BlockchainEvents};
+use sc_consensus::import_queue::ImportQueueService;
+use sc_network::{NetworkPeers, NetworkService};
+use sp_api::BlockT;
+use sp_blockchain::HeaderBackend;
+use std::future::Future;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+use subspace_networking::Node;
+use tracing::{info, trace};
+
+/// How much time to wait for new block to be imported before timing out and starting sync from DSN.
+const NO_IMPORTED_BLOCKS_TIMEOUT: Duration = Duration::from_secs(10 * 60);
+/// Frequency with which to check whether node is online or not
+const CHECK_ONLINE_STATUS_INTERVAL: Duration = Duration::from_secs(10);
+
+#[derive(Debug)]
+enum NotificationReason {
+    NoImportedBlocks,
+    WentOnlineSubspace,
+    WentOnlineSubstrate,
+}
+
+/// Create node observer that will track node state and send notifications to worker to start sync
+/// from DSN.
+pub(super) fn create_observer_and_worker<Block, Client>(
+    network_service: Arc<NetworkService<Block, <Block as BlockT>::Hash>>,
+    node: Node,
+    client: Arc<Client>,
+    mut import_queue_service: Box<dyn ImportQueueService<Block>>,
+) -> (
+    impl Future<Output = ()> + Send + 'static,
+    impl Future<Output = Result<(), sc_service::Error>> + Send + 'static,
+)
+where
+    Block: BlockT,
+    Client: HeaderBackend<Block>
+        + BlockBackend<Block>
+        + BlockchainEvents<Block>
+        + Send
+        + Sync
+        + 'static,
+{
+    let (tx, rx) = mpsc::channel(0);
+    let observer_fut = {
+        let node = node.clone();
+        let client = Arc::clone(&client);
+
+        async move { create_observer(network_service.as_ref(), &node, client.as_ref(), tx).await }
+    };
+    let worker_fut = async move {
+        create_worker(&node, client.as_ref(), import_queue_service.as_mut(), rx).await
+    };
+    (observer_fut, worker_fut)
+}
+
+async fn create_observer<Block, Client>(
+    network_service: &NetworkService<Block, <Block as BlockT>::Hash>,
+    node: &Node,
+    client: &Client,
+    notifications_sender: mpsc::Sender<NotificationReason>,
+) where
+    Block: BlockT,
+    Client: BlockchainEvents<Block> + Send + Sync + 'static,
+{
```

Comment:
A comment here would be useful to highlight this as an additional observer type.